File: 017_Letter_Combinations_of a_Phone_Number/017_1.py
from typing import List, Dict
from queue import Queue
from collections import deque, Counter
import math
from heapq import heappush, heappushpop

class Solution:
    def letterCombinations(self, digits: str) -> List[str]:
        if not digits:
            return []

        phone_map = {
            '2': "abc", 
            '3': "def", 
            '4': "ghi", 
            '5': "jkl", 
            '6': "mno", 
            "8": "tuv",
            "9": "wxyz",
        }
        res = []
        combination = ""
        def backtrack(index, comb):
            if index == len(digits):
                res.append(comb)
                return
            letters = phone_map[digits[index]]
            for letter in letters:
                backtrack(index + 1, comb + letter)
        backtrack(0, combination)
        return res


if __name__ == "__main__":
    s = Solution()
    print(s.letterCombinations("23"))

File: 017_Letter_Combinations_of a_Phone_Number/017_1.cpp
#include <iostream>
#include <string>
#include <vector>
#include <climits>
#include <algorithm>
#include <cstring>
#include <unordered_map>
#include <unordered_set>
#include <map>
#include <stack>
#include <bitset>
#include <array>
#include <queue>
#include <sstream>
#include <set>
using namespace std;

class Solution {
public:
    vector<string> letterCombinations(string digits) {
        if (digits.empty()) return {};
        unordered_map<char, string> map {
            {'2', "abc"},
            {'3', "def"},
            {'4', "ghi"},
            {'5', "jkl"},
            {'6', "mno"},
            {'7', "pqrs"},
            {'8', "tuv"},
            {'9', "wxyz"}
        };
        vector<string> res;
        string comb;
        backtrack(digits, 0, res, comb, map);
        return res;
    }

private:
    void backtrack(string& digits, int index, vector<string> &res,
            string &comb, unordered_map<char, string> map) {
        if (index == digits.size()) res.emplace_back(comb);
        else {
            char d = digits[index];
            string letters = map.at(d);
            for (const char &letter : letters) {
                comb.push_back(letter);
                backtrack(digits, index + 1, res, comb, map);
                comb.pop_back();
            }
        }
    }
};

int main() {
    Solution s;
    string str = "23";
    vector<string> res = s.letterCombinations(str);
    for (int i = 0; i < res.size(); i++) {
        cout << res[i] << " ";
    }
}


